### PR TITLE
Use OnceCell to store server versions

### PR DESCRIPTION
**Pull request by jsparber**
Since the server versions isn't modified we can use a OnceCell and
get rid of the Mutex.

### CODE REVIEW

**Comment by jsparber** [comment]
Obviously https://docs.rs/async-once-cell/latest/async_once_cell/struct.OnceCell.html#method.get_or_try_init doesn't have the same `fn` definition.

**Comment by jplatte** [comment]
Given that these have a fairly different API, let's not have this in `matrix-sdk-common` and instead import from the two different places in `matrix-sdk`.

**Comment by jsparber** [comment]
> Given that these have a fairly different API, let's not have this in `matrix-sdk-common` and instead import from the two different places in `matrix-sdk`.

that makes sense

**Comment by jsparber** [comment]
> This crate mentions this:
> 
> > This crate is extremely new and has therefore not extensively been tested. Next to that, the author is also inexperienced in working with unsafe Rust. While care has been taken in making this crate safe, it is not recommended for production use. As mentioned in the Contributing section: feedback and pull requests are very much appreciated, even more so if you see any problems with the soundness of the crate.
> 
> Do we really want to use this crate for something that should be relatively simple?

There are two crates with pretty much the same name:`async-once-cell` and `async-oncecell`. The first one is a fork of the other and it seams like the author is more confident about the crate.

File: crates/matrix-sdk/Cargo.toml [hunk 1]
```diff
@@ -118,6 +118,7 @@ optional = true
 
 [target.'cfg(target_arch = "wasm32")'.dependencies]
 wasm-timer = "0.2.5"
+async-once-cell = "0.3.0"
```
Comment (jplatte) [inline crates/matrix-sdk/Cargo.toml:121]:
```suggestion
async-once-cell = "0.3.0"
wasm-timer = "0.2.5"
```

File: crates/matrix-sdk/src/client/builder.rs [hunk 2]
```diff
@@ -324,16 +323,23 @@ impl ClientBuilder {
...
+        let server_versions = if let Some(server_versions) = self.server_versions {
+            let cell = OnceCell::new();
+            cell.get_or_init(async move { server_versions }).await;
+            cell
+        } else {
+            OnceCell::new()
         };
 
+        #[cfg(not(target_arch = "wasm32"))]
+        let server_versions = OnceCell::new_with(self.server_versions);
```
Comment (jplatte) [inline crates/matrix-sdk/src/client/builder.rs:336]:
```suggestion
        let server_versions = tokio::OnceCell::new_with(self.server_versions);
```

File: crates/matrix-sdk/src/client/builder.rs [hunk 3]
```diff
@@ -63,7 +66,7 @@ pub struct ClientBuilder {
     request_config: RequestConfig,
     respect_login_well_known: bool,
     appservice_mode: bool,
-    server_versions: Option<Arc<[MatrixVersion]>>,
+    server_versions: Option<Vec<MatrixVersion>>,
```
Comment (jplatte) [inline crates/matrix-sdk/src/client/builder.rs:69]: Switching from `Arc<[_]>` to `Vec<_>` and using a borrow in `HttpClient` makes the future returned by `Client::send` not `'static`, can you please revert that part?

File: crates/matrix-sdk/src/client/mod.rs [hunk 4]
```diff
@@ -1486,30 +1490,35 @@ impl Client {
...
+    async fn server_versions(&self) -> HttpResult<&[MatrixVersion]> {
+        #[cfg(target_arch = "wasm32")]
+        let server_versions =
+            self.inner.server_versions.get_or_try_init(self.request_server_versions()).await?;
+
+        #[cfg(not(target_arch = "wasm32"))]
+        let server_versions = self
+            .inner
+            .server_versions
+            .get_or_try_init(move || self.request_server_versions())
```
Comment (jplatte) [inline crates/matrix-sdk/src/client/mod.rs:1518]: This `move` keyword seems unnecessary.

File: crates/matrix-sdk/src/client/builder.rs (hunk 3 above)
Comment (jsparber) [reply crates/matrix-sdk/src/client/builder.rs:69]: Also for `Client::server_versions`?

File: crates/matrix-sdk/src/client/mod.rs [hunk 5]
```diff
@@ -1486,28 +1490,35 @@ impl Client {
         self.inner.http_client.send(request, config, self.server_versions().await?).await
     }
 
-    async fn server_versions(&self) -> HttpResult<Arc<[MatrixVersion]>> {
-        let mut server_versions = self.inner.server_versions.lock().await;
+    async fn request_server_versions(&self) -> HttpResult<Arc<[MatrixVersion]>> {
```
Comment (jplatte) [inline crates/matrix-sdk/src/client/mod.rs:1493]: You can inline this as an `async {}` block before the two `let server_versions` in the `server_versions` function.

Comment (jplatte) [reply crates/matrix-sdk/src/client/mod.rs:1493]: Other than that and CI still failing, this looks good now!

**Review by jplatte — APPROVED** [review]

**Comment by jplatte** [comment]
@poljar are you okay with the async-once-cell (wasm-only) dependency given Julian's later comment? I personally don't *love* it, but conceptually `OnceCell` is the right abstraction / `Mutex` is not and it's easy enough to add extra security measures like setting up notifications of GitHub releases and reviewing those for small deps like this (which I think we may want to do for other dependencies too, at some point).

File: crates/matrix-sdk/src/client/mod.rs [hunk 6]
```diff
@@ -66,6 +68,8 @@ use ruma::{
     ServerName, UInt,
 };
 use serde::de::DeserializeOwned;
+#[cfg(not(target_arch = "wasm32"))]
```
Comment (gnunicorn) [inline crates/matrix-sdk/src/client/mod.rs:71]: Note: this is one of the cases, where I'd rather have the two imports next to one another to make that cfg-relationship obvious than insisting on a alphabetical order...

**Review by gnunicorn — APPROVED** [review]

File: crates/matrix-sdk/src/client/mod.rs (hunk 6 above)
Comment (jsparber) [reply crates/matrix-sdk/src/client/mod.rs:71]: rustfmt resorted them.

**Pull request merged**